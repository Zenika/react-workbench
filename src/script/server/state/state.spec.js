/* eslint-env jest */

jest.mock('../db', () => {
  const read = jest.fn()
  const write = jest.fn()
  const list = jest.fn()

  return () => ({ read, write, list })
})

const db = require('../db')
const service = require('./state')

describe('server/model/state', () => {
  beforeEach(() => {
    db().read.mockReset()
    db().write.mockReset()
    db().list.mockReset()

    db().read.mockRestore()
    db().write.mockRestore()
    db().list.mockRestore()
  })

  describe('list', () => {
    it('should handle empty directory', async () => {
      // data & mocks
      db().list.mockImplementation(jest.fn(() => []))

      // calls
      const result = await service.list()()

      // asserts
      expect(result).toMatchSnapshot()
      expect(db().list.mock.calls).toMatchSnapshot()
    })

    it('should returns all files content', async () => {
      // data & mocks
      db().list.mockImplementation(jest.fn(() => ['file 1', 'file 2']))
      db().read
        .mockReturnValueOnce('content file 1')
        .mockReturnValueOnce('content file 2')

      // calls
      const result = await service.list()()

      // asserts
      expect(result).toMatchSnapshot()
      expect(db().list.mock.calls).toMatchSnapshot()
      expect(db().read.mock.calls).toMatchSnapshot()
    })
  })

  describe('create', () => {
    it('should append a new state', () => {
      // data
      const state = { a: 'state' }

      // calls
      service.create(state)()

      // asserts
      expect(db().write.mock.calls.length).toBe(1)
    })
  })

  describe('read', () => {
    it('should read states', async () => {
      // data
      const mockStates = ['state1', 'state2']

      // mocks
      db().read.mockImplementation(jest.fn(async () => mockStates))

      // calls
      const states = await service.read()()

      // asserts
      expect(db().read.mock.calls.length).toBe(1)
      expect(states).toBe(mockStates)
    })

    it('should initialize states', async () => {
      // calls
      const states = await service.read()()

      // asserts
      expect(db().read.mock.calls.length).toBe(1)
      expect(states).toEqual({})
    })
  })
})
