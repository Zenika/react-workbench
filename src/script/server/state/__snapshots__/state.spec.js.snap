// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`server/model/state list should handle empty directory 1`] = `Array []`;

exports[`server/model/state list should handle empty directory 2`] = `
Array [
  Array [],
]
`;

exports[`server/model/state list should returns all files content 1`] = `
Array [
  "content file 1",
  "content file 2",
]
`;

exports[`server/model/state list should returns all files content 2`] = `
Array [
  Array [],
]
`;

exports[`server/model/state list should returns all files content 3`] = `
Array [
  Array [
    "file 1",
  ],
  Array [
    "file 2",
  ],
]
`;
